libnmstate/state.py

```diff
@@ -45,13 +46,65 @@
 from libnmstate.schema import InterfaceState
 from libnmstate.schema import InterfaceType
 from libnmstate.schema import Route
+from libnmstate.schema import RouteRule
 
 
 NON_UP_STATES = (InterfaceState.DOWN, InterfaceState.ABSENT)
 
 
 @total_ordering
-class RouteEntry(object):
+class StateEntry:
+    def __init__(self):
+        pass
+
+    def keys(self):
+        """
+        Return the tuple representing this entry, could be used for hashing or
+        comparing.
+        """
+        return ()
+
+    def __hash__(self):
+        return hash(self.keys())
+
+    def __eq__(self, other):
+        return self is other or self.keys() == other.keys()
+
+    def __lt__(self, other):
+        return self.keys() < other.keys()
+
+    def __repr__(self):
+        return str(self.to_dict())
+
+    @property
+    def absent(self):
+        raise NmstateNotImplementedError(
```

Comment:
If this code is called, it should not be a user's fault by calling something that is not implemented. If it is a user-facing exception, the error message should be something like 'Removing route/rule entry ... not supported' to help the user figure out what they tried to do. Does this really happen? If everything should always implement this, it is a good candidate for an abstract method: https://docs.python.org/3/library/abc.html#abc.abstractmethod

Reply:
You can use here a simple `raise NotImplementedError`.
abc is also a candidate, possibly even better. It also limits the instantiation of the base class.
I am good with both, whatever works for you.

Reply:
User will never use this class. This is supposed to be internal only.
So I will raise `NmstateInternalError` for user to report a bug to use when happens.

Reply:
The `abc` is over kill for this, we may consider using it when more than 2 classes need this.

Reply:
Using `NotImplementedError` is a good compromise I think.
This is an abstract and should be treated as such by the developer.